### DISCUSSION TITLE
Inference only uses last token except in the first forward pass

### DISCUSSION THREAD

**Discussion by tom-huntington** [openai/whisper#565] [General]
**Edit**: Woops ignore this issue, this is just how key value caching is implemented.

https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/decoding.py#L141-L143

This is probably much more efficient. Although I'm surprised. I though whisper would be using the power autoregressive language models, but it doesn't.

So this must mean there is no control over where the timestamps happen.

They just get filtered out here

https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/transcribe.py#L195

Actually, you can just take the argmax of the the timestamp logits to get the timestamps for each word https://github.com/openai/whisper/discussions/3#discussioncomment-3730914

**Comment by tom-huntington** [comment]
https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/model.py#L75-L79

so after forward is called on `self.k` and `self.v` the following hook is called to append the previously cached results, just as the comment says.

https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/model.py#L254

The keys and values for the audio features are also cached, but a hook is not used for these.

I guess masking makes the caching possible
